### PR TITLE
feat: merge all lockups into a single contract

### PR DESCRIPTION
**Pull request by smol-ninja**
Closes https://github.com/sablier-labs/v2-core/issues/1074. 

#### Since the work is split into two PRs. The order of merging would be:
1. https://github.com/sablier-labs/v2-core/pull/1077
2. https://github.com/sablier-labs/v2-core/pull/1069

- [x] Rename `Family` to `Model`
- [x] Rename `SablierLockup` to `SablierLockupBase` and `SablierLockupOne` to `SablierLockup`
- [x] Emit separate events in each lockup create function
- [x] Include `endTime` in `CreateWithTimestamps` struct
- [x] Refactor `calculateStreamedAmountForSegments` to disallow unnecessary inputs
- [x] Set `Stream` using JSON in create functions
- [x] Update Helpers: split it into two libraries
- [x] Update Errors
- [x] Refactor calculation function in LD for readability
- [x] Update inherited components
- [x] Update shell scripts
- [x] Update precompiles
- [x] Update deployment scripts
- [x] Update periphery
- [x] Run a deployment on Sepolia. [etherscan](https://sepolia.etherscan.io/address/0x6fe81f4bf1af1b829f0e701647808f3aa4b0bde7#code)
- [x] Include public libraries in artifacts
- [x] prefix error names for libraries to SablierHelpers and SablierVestingMath

### CODE REVIEW

**Comment by smol-ninja** [comment]
@andreivladbrg let me know if you get a chance to review this approach. For the records, this is just to demonstrate the idea and does not confirm the actual changes that should be made.

**Comment by andreivladbrg** [comment]
@smol-ninja at your request, I am going to leave here some of the suggestion/things I would change/improve:

---

I believe we should have different `create` events emitted (similarly to `claim` on airstream). So that the unique
params related to the model used is emitted.

---

Add `endTime` in the `Lockup.CreateWithTimestamps` struct, allowing all create function signatures to follow this pattern: `function create(CommonParams calldata params, UniqueStorage memory unique) external override returns (uint256 streamId);`. This change would require us to check that the timestamp of the last segment/tranche matches the `endTime`.

Move this logic in `_calculateStreamedAmount` from individual `LD/LL/LT` functions (since it is shared logic).

```solidity
uint256 blockTimestamp = block.timestamp;

if (_streams[streamId].startTime <= blockTimestamp) {
    return 0;
}

if (block.timestamp >= _streams[streamId].endTime) {
    return _streams[streamId].amounts.deposited;
}
```

---

I would move `Helpers.checkAndCalculateBrokerFee` check in `_create`.

---

There is no need to pass the entire stream struct into `calculateStreamedAmountForSegments`.


---

We can keep the current approach to updating storage for the stream struct, as I find it more readable:

```solidity
_streams[streamId] = Lockup.Stream({
    amounts: Lockup.Amounts({ deposited: createAmounts.deposit, refunded: 0, withdrawn: 0 }),
    asset: params.asset,
    endTime: params.timestamps.end,
    isCancelable: params.cancelable,
    isDepleted: false,
    isStream: true,
    isTransferable: params.transferable,
    sender: params.sender,
    startTime: params.timestamps.start,
    wasCanceled: false
});
```

---

I am not a fan of the `Family` enum name, I would rather chose something like `StreamingModel`,
`StreamType`,`LockupModel` or `LockupType`.

---

The `Helpers` library becomes a long one and maybe hard to follow, what if we going to have two libraries:

1. Calculation library: for all the calculation functions
2. Checks (validations)

For the first one, we can place all calculation logic for all models in one place.

<details><summary>Click to see what I mean</summary>
<p>

```solidity
function _calculateStreamedAmount(uint256 streamId) internal view override returns (uint128 streamedAmount) {
    // If the end time is not in the future, return the deposited amount.
    uint256 blockTimestamp = block.timestamp;

    if (_streams[streamId].startTime <= blockTimestamp) {
        return 0;
    }

    if (block.timestamp >= _streams[streamId].endTime) {
        return _streams[streamId].amounts.deposited;
    }

    Family family = _streams[streamId].family;

    if (family == Lockup.Family.LOCKUP_LINEAR) {
        streamedAmount = Calculation.linearStream(params);
    } else if (family == Lockup.Family.LOCKUP_DYNAMIC) {
        streamedAmount = Calculation.calculateStreamedAmountForSegments(params);
    } else if (family == Lockup.Family.LOCKUP_TRANCHED) {
        streamedAmount = Calculation.sumOfPassedTranches(params);
    }
}
```

</p>
</details>

A major benefit of this is that we can create a mock for this library independently and isolate tests for calculation logic, which is arguably the most _fragile_ part of the code.

---

If possible, and assuming no significant downsides, I would divide the dynamic calculation function into two parts:

1. sum of passed segments
   (`function calculatePassedSegments(uint128[] memory amounts, uint40[] memory timestamps) internal pure returns (uint128)`)
   so that we can use it for both `LOCKUP_DYNAMIC` and `LOCKUP_TRANCHED` models.
2. current segment streamed amount

**Comment by smol-ninja** [comment]
> I would rename the current SablierLockup contract to SablierLockupBase and the SablierLockupOne to SablierLockup to match what we do in flow.

Agree. `SablierLockupOne` was just a placeholder name to avoid conflict with existing `SablierLockup` contract while writing that down.

> I believe we should have different create events emitted (similarly to claim on airstream). So that the unique
params related to the model used is emitted.

Unique params related to the model are still emitted through separate events. But your suggestion sounds good to me. If it increase the contract size by a good margin, would it be OK to have shared event and a separate lockup specific event? 

> Add endTime in the Lockup.CreateWithTimestamps struct, allowing all create function signatures to follow this pattern: function create(CommonParams calldata params, UniqueStorage memory unique) external override returns (uint256 streamId

Sounds good. Will experiment with this idea.

> I would move Helpers.checkAndCalculateBrokerFee check in _create

Why? This would increase contract size. 

> There is no need to pass the entire stream struct into calculateStreamedAmountForSegments

Sounds good.

> We can keep the current approach to updating storage for the stream struct

Agree.

> I am not a fan of the Family enum name, I would rather chose something like StreamingModel,
StreamType,LockupModel or LockupType.

"Fan" is a personal choice. My rationale is that (1) its a single word (2) Inspired from [Family of Curves](https://en.wikipedia.org/wiki/Family_of_curves) which means each curves in the family share the same formula which is indeed a case here. Therefore I find `Family` more suitable for this. This is also keyword that can be remembered easily.

> The Helpers library becomes a long one and maybe hard to follow

May be it is not hard to follow. We have bigger files than than in our repos. To increase readability, I can put categories in it.

> A major benefit of this is that we can create a mock for this library independently and isolate tests for calculation logic, which is arguably the most fragile part of the code

Sorry, I do not understand it. Can you please re-explain?

> If possible, and assuming no significant downsides, I would divide the dynamic calculation function

Sounds good. Will try this approach.

**Comment by PaulRBerg** [comment]
I would suggest `Model` or `DistributionModel` (not `StreamingModel` because `LockupTranched` is not streaming) because this is how we have been referring to the Linear, Dynamic, and Tranched categories all along. Just search for "model" in the `src` folder, and you will find it used in the NFT SVG and many explanatory comments.

**Comment by smol-ninja** [comment]
> because `LockupTranched` is not streaming

Why isn't `LockupTranched` categorised as streaming? If a stream is defined based on a per-second unlocks then even Lockup Dynamic is not streaming even though LD can be used to create per-second streams but then LT can also be used to  create per-second streams by setting tranche duration to 1s. 

IMO, streaming can be defined as the unlocking of tokens over a period of time (referring to [multiple cadences discussion](https://github.com/sablier-labs/flow/discussions/152)). So, as long as there are multiple unlocks periodically, it can be considered a stream.

**Comment by PaulRBerg** [comment]
We discussed the streaming terminology here: https://github.com/sablier-labs/v2-core/discussions/708

**Comment by smol-ninja** [comment]
Sounds great. Thanks for pointing me to the other discussion. I am in for `LockupModel`/`DistributionModel`.

**Comment by andreivladbrg** [comment]
re terminology , i like the simple `Model` 

---

> Unique params related to the model are still emitted through separate events. But your suggestion sounds good to me. If it increase the contract size by a good margin, would it be OK to have shared event and a separate lockup specific event?

hmm, didn't see them, i talking about: cliff for LL, segments for LD, and tranches for LT

> I would move Helpers.checkAndCalculateBrokerFee check in _create
Why? This would increase contract size

on contrary, this would decrease the size as there would only one time this function called (in the common `_create`). the version i reviewed was like this:

<details><summary>Click to see code</summary>
<p>

```solidity
function _createLD(Params memory params) internal returns(uint256) {
   Helpers.checkAndCalculateBrokerFee(params);
   // --snip--
   _create(params);
}
function _createLL(Params memory params) internal returns(uint256) {
   Helpers.checkAndCalculateBrokerFee(params);
   // --snip--
   _create(params);
}
function _createLD(Params memory params) internal returns(uint256) {
   Helpers.checkAndCalculateBrokerFee(params);
   // --snip--
   _create(params);
}
function _create(Params memory params) internal returns(uint256) {
   // --snip--
}

```

I am suggesting: 

```solidity
function _createLD(Params memory params) internal returns(uint256) {
   // --snip--
   _create(params);
}
function _createLL(Params memory params) internal returns(uint256) {
   // --snip--
   _create(params);
}
function _createLD(Params memory params) internal returns(uint256) {
   // --snip--
   _create(params);
}
function _create(Params memory params) internal returns(uint256) {
   Helpers.checkAndCalculateBrokerFee(params);
}
```

</p>
</details> 

---

> Sorry, I do not understand it. Can you please re-explain?

My idea was to divide the logic as follows:
1. Conditional checks (e.g. if the start is in the future, return 0; if the end is in the past, return the deposit) within the contract itself.
2. Calculation logic isolated in a library (e.g. calculation of: elapsed time, elapsed percentage, streamed amount, etc.)

This structure would improve organization and allow testing to be divided between the two parts.

```mermaid
flowchart TD
    subgraph SmartContractSystem
        direction TB
        A[Contract A]
        B(Library Calculation Logic)
        C(Mock Library for Testing)
        A -->|Check: if X then Y, if W then Z| ACheck[Conditional Checks]
    end
    
    %% Contract A structure
    
    %% A -->|Calls| B[Library Calculation Logic]

    %% Testing Structure
    subgraph Testing
        direction TB
        T[Test Calculation Logic]
        T1[Test Conditional Checks]
        T1 --> |Uses| ACheck
        T -->|Uses| C
    end

    %% Explanation
    ACheck -->|If Conditions Met| B
    C -.->|Mock Contract| B
```

**Comment by smol-ninja** [comment]
> I would move Helpers.checkAndCalculateBrokerFee check in _create

Its not possible to do so. The reason is that the returned value of `checkAndCalculateBrokerFee`: `createAmounts` is used in the lockup specific `check` functions. If you have a better solution, please let me know or feel free to push a commit directly.

----------

@andreivladbrg I have included all your suggestions including the library split. They were all good. I have also updated the core source contracts. Feel free to review the `src/core` directory as the contract work is done from my side. You can also see whats done and whats remaining in the PR description.

The size of the `SablierLockup` contract is `24,284` at 1000 runs which is `292` less than the allowed.

**Comment by smol-ninja** [comment]
@andreivladbrg this PR is ready for your review. A separate PR is opened to implement the tests and the remaining changes (https://github.com/sablier-labs/v2-core/pull/1077). Please see the description in both the PRs to see what they cover.

File: src/core/LockupNFTDescriptor.sol [hunk 1]
```diff
@@ -7,7 +7,7 @@ import { IERC721Metadata } from "@openzeppelin/contracts/token/ERC721/extensions
 import { Base64 } from "@openzeppelin/contracts/utils/Base64.sol";
 import { Strings } from "@openzeppelin/contracts/utils/Strings.sol";
 import { ILockupNFTDescriptor } from "./interfaces/ILockupNFTDescriptor.sol";
-import { ISablierLockup } from "./interfaces/ISablierLockup.sol";
+import { ISablierLockupBase } from "./interfaces/ISablierLockupBase.sol";
```
Comment (andreivladbrg) [inline src/core/LockupNFTDescriptor.sol:10]: we can use `ISablierLockup` instead of `ISablierLockupBase`

File: src/core/abstracts/SablierLockupBase.sol [hunk 2]
```diff
@@ -31,13 +31,13 @@ abstract contract SablierLockup is
                                   STATE VARIABLES
     //////////////////////////////////////////////////////////////////////////*/
 
-    /// @inheritdoc ISablierLockup
+    /// @inheritdoc ISablierLockupBase
     UD60x18 public constant override MAX_BROKER_FEE = UD60x18.wrap(0.1e18);
```
[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/abstracts/SablierLockupBase.sol:35]: These constants are specific to Lockup contracts. For example, we may have a lockup contract with a different max fee. Therefore, they should be a part of the this contract and not the libraries. Another benefit is that it makes libraries re-usable across future release of lockups.

File: src/core/types/DataTypes.sol [hunk 3]
```diff
@@ -96,63 +150,25 @@ library Lockup {
...
-        uint40 startTime;
-        Segment[] segments;
-        Broker broker;
+    /// @notice Struct encapsulating the Lockup timestamps.
+    /// @param start The Unix timestamp for the stream's start.
+    /// @param cliff The Unix timestamp for the cliff period's end. A value of zero means there is no cliff.
+    /// @param end The Unix timestamp for the stream's end.
+    struct Timestamps {
+        uint40 start;
+        uint40 cliff;
```
[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/types/DataTypes.sol:164]: Sorry I did not understand your suggestion here. 

>  I think it doesn't do justice to keep it here

Where should it be kept?

> what is the reason not to include all timestamps in segments/tranches

For example? All segments have a timestamp which represents the end time of that segment.

File: src/core/SablierLockup.sol [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+        MAX_COUNT = maxCount;
+        nextStreamId = 1;
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                           USER-FACING CONSTANT FUNCTIONS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc ISablierLockup
+    function getCliff(uint256 streamId) external view override notNull(streamId) returns (uint40 cliff) {
```
[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/SablierLockup.sol]: I thought about it. But then in `getTimestamps`, we return `cliff`. So it did not make sense for me to revert on `getCliff` but return `cliff `in `getTimestamps`. Otherwise I agree with you. Whats your thought on this?

File: src/core/SablierLockup.sol [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+        params.asset.safeTransferFrom({ from: msg.sender, to: address(this), value: createAmounts.deposit });
+
+        // Interaction: pay the broker fee, if not zero.
+        if (createAmounts.brokerFee > 0) {
+            params.asset.safeTransferFrom({ from: msg.sender, to: params.broker.account, value: createAmounts.brokerFee });
+        }
+
+        unchecked {
+            // Effect: bump the next stream ID.
+            nextStreamId = streamId + 1;
```
[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/SablierLockup.sol:365]: Sounds good.

File: src/core/libraries/VestingMath.sol [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+            unchecked {
+                streamedAmount += tranches[i].amount;
+            }
+        }
+
+        return streamedAmount;
+    }
+
+    /// @dev Calculate the timestamps and return the segments.
+    function calculateSegmentTimestamps(LockupDynamic.SegmentWithDuration[] memory segments)
```
[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/libraries/VestingMath.sol:182]: Why? My rationale was that since its a calculation function so all calculation function in vesting math and check function in helpers.

File: src/core/SablierLockup.sol [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+        // Checks, Effects and Interactions: create the stream.
+        streamId = _createLT(params, tranches);
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                           INTERNAL CONSTANT FUNCTIONS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc SablierLockupBase
+    function _calculateStreamedAmount(uint256 streamId) internal view override returns (uint128 streamedAmount) {
```
[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/SablierLockup.sol:274]: Good catch. Also, `function _calculateStreamedAmount` should be removed from `SablierLockupBase`. Right?

File: src/core/SablierLockup.sol [hunk 8]
```diff
@@ -0,0 +1,508 @@
...
+        // If the start time is in the future, return zero.
+        uint40 blockTimestamp = uint40(block.timestamp);
+        if (_streams[streamId].startTime >= blockTimestamp) {
```
[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/SablierLockup.sol:277]: We can but it does not matter. Its only called once in every if condition.

File: src/core/libraries/Helpers.sol [hunk 9]
```diff
@@ -189,6 +193,13 @@ library Helpers {
             revert Errors.SablierLockup_StartTimeNotLessThanFirstSegmentTimestamp(startTime, segments[0].timestamp);
         }
 
+        // Check: the end time equals the last segment's timestamp.
+        if (endTime != segments[segments.length - 1].timestamp) {
```
[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/libraries/Helpers.sol:197]: Agree.

File: src/core/SablierLockup.sol [hunk 10]
```diff
@@ -0,0 +1,496 @@
...
+            return VestingMath.calculateLockupDynamicStreamedAmount({
+                segments: _segments[streamId],
+                withdrawnAmount: _streams[streamId].amounts.withdrawn,
+                startTime: _streams[streamId].startTime
+            });
+        }
+        // Calculate streamed amount for Lockup Linear models.
+        else if (_streams[streamId].lockupModel == Lockup.Model.LOCKUP_LINEAR) {
+            // If the cliff time is in the future, return zero.
+            if (_cliffs[streamId] > uint40(block.timestamp)) {
```
[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/SablierLockup.sol:298]: Good point. Also its the role of `vesgtingMath` library to return the vested amount which in this condition is 0.

File: src/core/SablierLockup.sol [hunk 11]
```diff
@@ -0,0 +1,496 @@
...
+                           INTERNAL NON-CONSTANT FUNCTIONS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    function _create(
+        Lockup.CreateWithTimestamps memory params,
+        Lockup.CreateAmounts memory createAmounts,
+        Lockup.Model lockupModel
+    )
+        internal
+        returns (uint256 streamId)
```
[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/SablierLockup.sol:332]: Makes sense. Agree.

File: shell/update-counts.sh
[the remark this answers is not in the document]
Comment (smol-ninja) [reply shell/update-counts.sh:1]: Updated in [test PR](https://github.com/sablier-labs/v2-core/pull/1077). Its a part of that PR's todo.

File: src/core/libraries/Helpers.sol [hunk 12]
```diff
@@ -107,101 +121,29 @@ library Helpers {
...
-            revert Errors.SablierLockup_DepositAmountZero();
-        }
-
-        // Check: the start time is not zero.
-        if (timestamps.start == 0) {
-            revert Errors.SablierLockup_StartTimeZero();
-        }
-
+    /// @dev Checks the user-provided cliff and end times of a lockup linear stream.
+    function _checkCliffAndEndTime(uint40 startTime, uint40 cliffTime, uint40 endTime) private pure {
```
[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/libraries/Helpers.sol:125]: Agree. The reason I kept it as a separate function is to make all `checkCreateLockup` functions look similar. However, if you recommend this, then doesn't the same argument applies to `_checkSegments` and `_checkTranches` as well?

File: src/core/LockupNFTDescriptor.sol (hunk 1 above)
Comment (smol-ninja) [reply src/core/LockupNFTDescriptor.sol:10]: Agree.

File: src/core/types/DataTypes.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (andreivladbrg) [reply src/core/types/DataTypes.sol:164]: the comment was left on the `cliff` var

i am saying it shouldn't be included in the `Timestamp` struct

> For example? All segments have a timestamp which represents the end time of that segment.

yes, which are not included in this struct

---

we already talked about this on slack, and on [the discussion](https://github.com/sablier-labs/v2-core/discussions/1064#discussioncomment-11166968), you can nevermind this comment, as it was left before

File: src/core/SablierLockup.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (andreivladbrg) [reply src/core/SablierLockup.sol]: we decided not to include the cliff in timestamps, so we should revert

[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/SablierLockup.sol]: Agree.

File: src/core/libraries/VestingMath.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (andreivladbrg) [reply src/core/libraries/VestingMath.sol:182]: my rational is:

1. everything in `Helpers` is used pre stream creation
2. everything in `VestingMath` is used post stream creation

wdyt?

> calculation function so all calculation function in vesting math

yes, they are calculation function, but the lib is called "VestingMath" - so IMO it's not a part of this subdomain

[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/libraries/VestingMath.sol:182]: Interesting idea. It makes sense to me. Anyways, `calculateSegmentTimestamps` and `calculateTrancheTimestamps` convert one datatype to another so they ain't fit to the vesting math category. Happy to accept your suggestion.

File: src/core/libraries/Helpers.sol (hunk 12 above)
[the remark this answers is not in the document]
Comment (andreivladbrg) [reply src/core/libraries/Helpers.sol:125]: IIRC the `_checkSegments` and `_checkTranches` functions have also the role of preventing the stack too deep error

so i think we should keep them

[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/libraries/Helpers.sol:125]: Then for consistency, what do you think about checkCliffAndEndTime function?

File: src/core/abstracts/SablierLockupBase.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (andreivladbrg) [reply src/core/abstracts/SablierLockupBase.sol:35]: >These constants are specific to Lockup contracts

Max count, yes, since it can have different values accros chains, but why the max broker fee also?

> For example, we may have a lockup contract with a different max fee.

It is a `constant` and not `immutable`, which means a new contract must also be implemented declaring the `MAX_BROKER_FEE` with a new value. 

IMO, it is very unlikely to have different values for max broker fees; otherwise, we would have declared it as _immutable_.


The only purpose of this constant is to pass it in the Helpers.check() functions.


---

q: if this constant is declared in the library, would it be visible in Etherscan under the "Read Contract" category?

File: src/core/SablierLockup.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (andreivladbrg) [reply src/core/SablierLockup.sol:277]: If the calculation occurs for either `LL` or `LD` models, the `_streams[streamId].startTime` variable is read twice from storage (more gas)

File: src/core/libraries/Helpers.sol (hunk 12 above)
[the remark this answers is not in the document]
Comment (andreivladbrg) [reply src/core/libraries/Helpers.sol:125]: But it is not a requirement for `checkCliffAndEndTime` as for the other two, but if you have a strong opinion on this, feel free to keep it this way

File: src/core/abstracts/SablierLockupBase.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/abstracts/SablierLockupBase.sol:35]: >  it is very unlikely to have different values for max broker fees

I do not know why its not immutable but a constant. But my point is that Max Broker Fee and Max Protocol Fee are something that are specific to the protocol. On the other hand, libraries contain view only functions that take in some values, do some computations and return a new value.

Also, a user will be interested into getting max broker value from the lockup contract instead of going to the library address and checking the max value there. Also libraries are not allowed to have state variables and a constant is also a state variable (an immutable one). So imo it goes against the definition of a library.

> if this constant is declared in the library, would it be visible in Etherscan under the "Read Contract" category?

Should be.

File: src/core/SablierLockup.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/SablierLockup.sol:277]: You are right. I did not see that.

File: src/core/libraries/Helpers.sol (hunk 12 above)
[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/libraries/Helpers.sol:125]: I agree with what you are saying. It just looks beautiful this way. If you dont have a problem, lets keep it this way.

Screenshot 2024-11-08 at 17 16 26

File: src/core/abstracts/SablierLockupBase.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/abstracts/SablierLockupBase.sol:35]: @andreivladbrg this is the last comment thats awaiting your reply.

[the remark this answers is not in the document]
Comment (andreivladbrg) [reply src/core/abstracts/SablierLockupBase.sol:35]: >I do not know why its not immutable but a constant. 

it is more a business decision. but IMO practically i don't see a scenario where we would want a fee > 10%, so we can keep it as constant

> But my point is that Max Broker Fee and Max Protocol Fee are something that are specific to the protocol.

yeah, fair point

> On the other hand, libraries contain view only functions that take in some values, do some computations and return a new value.

yes, they check against a value passed against a constant

also, having constants in libraries is a common thing: 

https://github.com/Uniswap/v3-core/blob/d8b1c635c275d2a9450bd6a78f3fa2484fef73eb/contracts/libraries/TickMath.sol#L8-L16


> Also libraries are not allowed to have state variables and a constant is also a state variable (an immutable one). So imo it goes against the definition of a library.

constant and immutable are a part of the bytecode, no matter where that is placed (not in storage). so libraries are allowed to have "state variables". 

---

this comment: "Also, a user will be interested into getting max broker value from the lockup contract instead of going to the library address and checking the max value there"

and this one: "Should be."

i think they are in contradictory.

---

since it is more a business decision, and it's something specific to the protocol let's keep it here

[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/core/abstracts/SablierLockupBase.sol:35]: My arguments were bad. I was wrong to say that libraries can't have constant whereas they can.

I was against having broker fee constant in `VestingMath` library instead of Lockup contract because of the reason that a broker fee is more relevant to the lockup protocol than the library itself. So its about the placement of the variable. An example is `MAX_COUNT` which is placed in in `SablierLockup` and not in `SablierLockupBase` because its more appropriate for that contract since the segment length is compared against it.

> and this one: "Should be."

I thought your question was if Etherscan would show this value under "Read contract" category on the `VestingMath` library page for which I said "should be". Now I think you meant if it would show up on the Lockup contract page for which I would say "I don't know". Since the library would be deployed at a separate address, the Lockup contract might not show the constant value from the library.

**Comment by PaulRBerg** [comment]
See https://github.com/sablier-labs/v2-core/discussions/1082 before merging

**Comment by andreivladbrg** [comment]
> See #1082 before merging

I'd recommend merging this as is, without change the name, because the PR is already huge.

A different PR IMO would be better, also for `CHANGELOG`:
1. **Breaking:** Merge all lockup models into a single contract
2. **Breaking:** Rename Lockup to Vesing

**Comment by smol-ninja** [comment]
I agree with @andreivladbrg.

**Comment by PaulRBerg** [comment]
Agree.

File: src/core/types/DataTypes.sol [hunk 13]
```diff
@@ -52,6 +50,58 @@ library Lockup {
...
+        Timestamps timestamps;
+        Broker broker;
+    }
+
+    /// @notice Enum representing the different distribution models used to create lockup streams.
+    /// @dev These distribution models determine the vesting function used in the calculations of the unlocked assets.
+    enum Model {
+        LOCKUP_LINEAR,
+        LOCKUP_DYNAMIC,
+        LOCKUP_TRANCHED
```
Comment (andreivladbrg) [inline src/core/types/DataTypes.sol:102]: Since this enum is a part of `Lockup` namespace, should we rename the values to:

```solidity
library Lockup {
    enum Model {
        LINEAR,
        DYNAMIC,
        TRANCHED
    }
}
```

Comment (smol-ninja) [reply src/core/types/DataTypes.sol:102]: The return value won't have `LOCKUP` then.

Comment (andreivladbrg) [reply src/core/types/DataTypes.sol:102]: what do you mean?
the return value is a number (enums are `uint8`)

Comment (smol-ninja) [reply src/core/types/DataTypes.sol:102]: The benefit is that for integrators, the return value can helps them distinguish between LOCKUP_LINEAR and some other LINEAR model, from the perspective of aggregators. That was the reason I prefixed them with `LOCKUP`

Comment (smol-ninja) [reply src/core/types/DataTypes.sol:102]: > the return value is a number

Good point. Another thought I have is that on etherscan, if someone reads the function (or through RPC) and would see returned value as:
1. `LINEAR`
2. `LOCKUP_LINEAR`

Which one would be nice to read? I thought it would be nice to read `LOCKUP_LINEAR`, over `LINEAR`, what do you think?

Comment (andreivladbrg) [reply src/core/types/DataTypes.sol:102]: correct, for an integrators it looks better:

```solidity
import { Lockup } from "@sablier-labs/lockup/src/core/types/DataTypes.sol";

contract MyContract {
    // -- some logic ---
    function foo() public {
        Lockup.Model lockupModel = getLockupModel(id);
        if(lockupModel == LINEAR) {
            // do something
        }
        
        // vs
        if(lockupModel == LOCKUP_LINEAR) {
            // do something
        }
    }   
}
```
